auth/authenticator_test.go

```diff
@@ -0,0 +1,253 @@
+		url, username, password := "http://couch", "user", "pass"
+		auth, err := NewCouchDbSessionAuthenticator(url, username, password)
+		Expect(err).To(BeNil())
+		Expect(auth).ToNot(BeNil())
+		Expect(auth.AuthenticationType()).To(Equal(AUTHTYPE_COUCHDB_SESSION))
+
+		auth, err = NewCouchDbSessionAuthenticatorFromMap(map[string]string{
+			"URL":      url,
+			"USERNAME": username,
+			"PASSWORD": password,
+		})
+		Expect(err).To(BeNil())
+		Expect(auth).ToNot(BeNil())
+		Expect(auth.AuthenticationType()).To(Equal(AUTHTYPE_COUCHDB_SESSION))
+	})
+
+	It("Test Authenticator instantiation failures", func() {
+		errortests := []struct {
+			url, user, password string
+		}{
+			{"", "user", "password"},
+			{"localhost", "user", "password"},
+			{"https://localhost", "", "password"},
+			{"https://localhost", "{invalid-user}", "password"},
+			{"https://localhost", "user", ""},
+			{"https://localhost", "user", "{invalid-password}"},
+		}
+
+		for _, tt := range errortests {
+			_, err := NewCouchDbSessionAuthenticator(tt.url, tt.user, tt.password)
+			Expect(err).To(HaveOccurred())
+
+			_, err = NewCouchDbSessionAuthenticatorFromMap(map[string]string{
+				"URL":      tt.url,
+				"USERNAME": tt.user,
+				"PASSWORD": tt.password,
+			})
+			Expect(err).To(HaveOccurred())
+		}
+	})
+
+	Context("with standard test server", func() {
+		var (
+			server  *httptest.Server
+			request *http.Request
+			auth    *CouchDbSessionAuthenticator
+			err     error
+		)
+
+		BeforeEach(func() {
+			callNumber := 0
+			server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				callNumber++
+				cookie := &http.Cookie{
+					Name:  "AuthSession",
+					Value: fmt.Sprintf("fakefake-%d", callNumber),
+					Expires: time.Now().
+						Add(24*time.Hour - time.Duration(callNumber)*time.Minute),
+				}
+				http.SetCookie(w, cookie)
+				w.WriteHeader(http.StatusOK)
+			}))
+
+			builder, err := core.NewRequestBuilder(core.GET).
+				ResolveRequestURL("https://localhost", "/db", nil)
+			Expect(err).To(BeNil())
+
+			request, err = builder.Build()
+			Expect(err).To(BeNil())
+			Expect(request).ToNot(BeNil())
+
+			auth, err = NewCouchDbSessionAuthenticator(server.URL, "user", "pass")
+			Expect(err).To(BeNil())
+			Expect(auth).ToNot(BeNil())
+			Expect(auth.AuthenticationType()).To(Equal(AUTHTYPE_COUCHDB_SESSION))
+		})
+
+		AfterEach(func() {
+			server.Close()
+		})
+
+		It("Test setting custom http client on Authenticator", func() {
+			// set http.Client with custom timeout
+			auth.Client = &http.Client{Timeout: time.Second}
+
+			// call to Authenticate sets authenticator's client
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			Expect(auth.Client.Timeout).To(Equal(time.Second))
+		})
+
+		It("Test authentication re-request after expiration", func(done Done) {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			cookie, err := auth.getCookie()
```

Comment:
This is the first fetch from the cache, better to assert this, like from line 154 till 159. After you make this change the second fetch and its checks are not needed anymore.

Reply:
Those two fetches here to show that cache works as cache when session's not expired, without it the test might give false positive in case cache fetches cookie every time it's called.

Reply:
Then I would like to request a comment above this line that this is the first cache check, and an update to the second attempt, that is the second request from the cache.

Reply:
yeah, absolutely, it's evident it's too subtle and warrants the comment